### DISCUSSION TITLE
no NDK aarch64-linux-android-gcc on $PATH at (eval 12) line 142.

### DISCUSSION THREAD

**Discussion by dasanirban82** [openssl/openssl#26160] [Q&A]
Hi,
I am trying to build openssl 3.4 for android on a windows host, when building for windows there is no issue but when I try for Android I get below error :
I tried on VS Developer Command Prompt,PowerShell and on WSL Ubantu.The error reamins same.I want to build the android binaries,if windows host doesn't work then what should be the alternative for it?
C:\ANIDA\OpenSSL_v3\openssl>_set ANDROID_NDK_ROOT=C:\ANIDA\OpenSSL_v3\android-ndk-r27c_

C:\ANIDA\OpenSSL_v3\openssl>_set PATH=%ANDROID_NDK_ROOT%\toolchains\llvm\prebuilt\linux-x86_64\bin;%PATH%_

C:\ANIDA\OpenSSL_v3\openssl>_perl Configure android-arm64 --prefix=C:\ANIDA\OpenSSL_v3\openssl_libs enable-legacy -D__ANDROID_API__=29_
Configuring OpenSSL version 3.0.9 for target android-arm64
Using os-specific seed configuration

Failure!  build file wasn't produced.
Please read INSTALL.md and associated NOTES-* files.  You may also have to
look over your available compiler tool chain or change your configuration.

**no NDK aarch64-linux-android-gcc on $PATH at (eval 12) line 142.**

**C:\ANIDA\OpenSSL_v3\openssl>_perl configdata.pm --dump**_

Command line (with current working directory = .):

    C:\Strawberry\perl\bin\perl.exe Configure CC=clang CXX=clang++ --target=aarch64-linux-android enable-legacy

Perl information:

    C:\Strawberry\perl\bin\perl.exe
    5.40.0 for MSWin32-x64-multi-thread

Enabled features:

    aria
    asm
    async
    autoalginit
    autoerrinit
    autoload-config
    bf
    blake2
    bulk
    cached-fetch
    camellia
    capieng
    cast
    chacha
    cmac
    cmp
    cms
    comp
    ct
    deprecated
    des
    dgram
    dh
    dsa
    dso
    dtls
    dynamic-engine
    ec
    ec2m
    ecdh
    ecdsa
    engine
    err
    filenames
    gost
    idea
    legacy
    loadereng
    makedepend
    md4
    mdc2
    module
    multiblock
     nextprotoneg
    ocb
    ocsp
    padlockeng
    pic
    pinshared
    poly1305
    posix-io
    psk
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    secure-memory
    seed
    shared
    siphash
    siv
    sm2
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    ssl-trace
    static-engine
     stdio
    tests
    threads
    tls
    ts
    ui-console
    uplink
    whirlpool
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method

Disabled features:

    acvp-tests          [cascade]   OPENSSL_NO_ACVP_TESTS
    afalgeng            [not-linux] OPENSSL_NO_AFALGENG
    asan                [default]   OPENSSL_NO_ASAN
    buildtest-c++       [default]
    crypto-mdebug       [default]   OPENSSL_NO_CRYPTO_MDEBUG
    devcryptoeng        [default]   OPENSSL_NO_DEVCRYPTOENG
    ec_nistp_64_gcc_128 [default]   OPENSSL_NO_EC_NISTP_64_GCC_128
    egd                 [default]   OPENSSL_
     external-tests      [default]   OPENSSL_NO_EXTERNAL_TESTS
    fips                [default]
    fips-securitychecks [cascade]   OPENSSL_NO_FIPS_SECURITYCHECKS
    fuzz-afl            [default]   OPENSSL_NO_FUZZ_AFL
    fuzz-libfuzzer      [default]   OPENSSL_NO_FUZZ_LIBFUZZER
    ktls                [default]   OPENSSL_NO_KTLS
    md2                 [default]   OPENSSL_NO_MD2 (skip crypto\md2)
    msan                [default]   OPENSSL_NO_MSAN
    rc5                 [default]   OPENSSL_NO_RC5 (skip crypto\rc5)
    sctp                [default]   OPENSSL_NO_SCTP
    trace               [default]   OPENSSL_NO_TRACE
    ubsan               [default]   OPENSSL_NO_UBSAN
    unit-test           [default]   OPENSSL_NO_UNIT_TEST
    weak-ssl-ciphers    [default]   OPENSSL_NO_WEAK_SSL_CIPHERS
    zlib                [default]
    zlib-dynamic        [default]
    ssl3                [default]   OPENSSL_NO_SSL3
    ssl3-method         [default]   OPENSSL_NO_SSL3_METHOD

Config target attributes:

    AR => "lib",
    ARFLAGS => "/nologo",
    AS => "nasm",
    ASFLAGS => "-g",
    CC => "cl",
    CFLAGS => "/W3 /wd4090 /nologo /O2",
    CPP => "\$(CC) /EP /C",
    HASHBANGPERL => "/usr/bin/env perl",
    LD => "link",
    LDFLAGS => "/nologo /debug",
    MT => "mt",
    MTFLAGS => "-nologo",
     RANLIB => "CODE(0x2dfa3e87880)",
    RC => "rc",
    ar_resp_delim => "
",
    aroutflag => "/out:",
    asflags => "-Ox -f win64 -DNEAR",
    asm_arch => "x86_64",
    asoutflag => "-o ",
    bin_cflags => "/Zi /Fdapp.pdb",
    bin_lflags => "setargv.obj /subsystem:console /opt:ref",
    bn_ops => "SIXTY_FOUR_BIT",
    build_file => "makefile",
    build_scheme => [ "unified", "windows", "VC-common" ],
    cflags => "/Gs0 /GF /Gy /MD",
    coutflag => "/Fo",
    cppflags => "",
    defines => [ "OPENSSL_BUILDING_OPENSSL", "OPENSSL_SYS_WIN32", "WIN32_LEAN_AND_MEAN", "UNICODE", "_UNICODE", "_CRT_SECURE_NO_DEPRECATE", "_WINSOCK_DEPRECATED_NO_WARNINGS" ],
    disable => [  ],
    dso_cflags => "/Zi /Fddso.pdb",
    dso_scheme => "win32",
    enable => [  ],
    ex_libs => "ws2_32.lib gdi32.lib advapi32.lib crypt32.lib user32.lib",
    includes => [  ],
    ld_implib_flag => "/implib:",
    ld_resp_delim => "
",
    lddefflag => "/def:",
    ldoutflag => "/out:",
    ldpostoutflag => "",
    ldresflag => " ",
    lflags => "",
    lib_cflags => "/Zi /Fdossl_static.pdb",
    lib_cppflags => "",
     lib_defines => [ "L_ENDIAN" ],
    makedep_scheme => "VC",
    makedepcmd => "\$(CC) /Zs /showIncludes",
    module_cflags => "",
    module_cxxflags => undef,
    module_ldflags => "/dll",
    mtinflag => "-manifest ",
    mtoutflag => "-outputresource:",
    multilib => "-x64",
    perl_platform => "Windows::MSVC",
    perlasm_scheme => "auto",
    rcoutflag => "/fo",
    shared_cflag => "",
    shared_defflag => "",
    shared_defines => [  ],
    shared_ldflag => "/dll",
    shared_rcflag => "",
    shared_target => "win-shared",
    sys_id => "WIN64A",
    thread_defines => [  ],
    thread_scheme => "winthreads",
    unistd => "<unistd.h>",
    uplink_arch => "x86_64",

Recorded environment:

    AR =
    BUILDFILE =
    CC = C:\ANIDA\Work\Intune\android-ndk-r27c-linux\android-ndk-r27c\toolchains\llvm\prebuilt\linux-x86_64\bin\clang
    CFLAGS =
    CPPFLAGS =
    CROSS_COMPILE =
    CXX =
    CXXFLAGS =
    HASHBANGPERL =
    LDFLAGS =
    LDLIBS =
    OPENSSL_LOCAL_CONFIG_DIR =
    PERL =
    RANLIB =
    RC =
    RCFLAGS =
    WINDRES =
    __CNF_CFLAGS =
    __CNF_CPPDEFINES =
    __CNF_CPPFLAGS =
    __CNF_CPPINCLUDES =
    __CNF_CXXFLAGS =
    __CNF_LDFLAGS =
    __CNF_LDLIBS =

Makevars:

    AR              = lib
    ARFLAGS         = /nologo
    AS              = nasm
    ASFLAGS         = -g
    CC              = clang
    CFLAGS          = /W3 /wd4090 /nologo /O2 --target=aarch64-linux-android
    CPP             = $(CC) /EP /C
    CPPDEFINES      =
    CPPFLAGS        =
    CPPINCLUDES     =
    CXX             = clang++
     CXXFLAGS        = --target=aarch64-linux-android
    HASHBANGPERL    = /usr/bin/env perl
    LD              = link
    LDFLAGS         = /nologo /debug
    LDLIBS          =
    MT              = mt
    MTFLAGS         = -nologo
    PERL            = C:\Strawberry\perl\bin\perl.exe
    RANLIB          = ranlib
    RC              = rc
    RCFLAGS         =

NOTE: These variables only represent the configuration view.  The build file
template may have processed these variables further, please have a look at the
build file for more exact data:
    makefile

build file:

    makefile

build file templates:

    Configurations\common0.tmpl
    Configurations\windows-makefile.tmpl

**Comment by tom-cosgrove-arm** [comment]
I don't know enough to be able to help with Android generally, but there is a resolved issue #8941 where someone seemed to have the same problem.

Also, it's more normal to run `Configure` with just the target, e.g.

```
perl Configure android-arm64
```

(rather than using `--target=`). I'd expect that for such a specific build as Android the fewer command line options given the better.

**Reply by dasanirban82** [reply]
In the above resolved issue they tried in WSL Ubuntu, which also failed for me .It seems not able to find clang in the PATH.
anida_unix1@ANIDA-LAPTOP-HP:~/openssl$ ./Configure android-x86_64 --prefix=$HOME/openssl-android-x86_64
Configuring OpenSSL version 3.0.9 for target android-x86_64
Using os-specific seed configuration

Failure!  build file wasn't produced.
Please read INSTALL.md and associated NOTES-* files.  You may also have to
look over your available compiler tool chain or change your configuration.

no NDK x86_64-linux-android-gcc on $PATH at (eval 10) line 142.

**Comment by tom-cosgrove-arm** [comment]
(and this should probably be a discussion rather than an issue)

**Comment by Denzy7** [comment]
I tried solution in #8941 but didn't work.
I was able to compile 3.6-dev with clang albeit with some hacks:
- in `Configurations/15-android.conf`, remove `-mandroid` part so `$cflags` an empty string
- gcc and clang are very similar so in `$ANDROID_NDK_ROOT/toolchains/llvm/prebuilt/<os>-<arch>/bin`, just symlink `armv7a-linux-androideabi21-clang ==> arm-linux-androideabi-gcc` (api 21 compiled with no issues and is common. Also symlink for your target architecture like arm64 or x86)
- also symlink `llvm-ar ==> arm-linux-androideabi-ar` and `llvm-ranlib ==> arm-linux-androideabi-ranlib` (match your intended architecture)
- dont define `-D__ANDROID_API=21` at it's already defined

When testing in termux use `LD_LIBRARY_PATH` or `patchelf` rpath so it doesn't load `/system/lib/libssl.so`. I believe when using it in an apk it should prioritize the `libssl.so` in your apk

I tested on ndk 23.1.7779620 


[EDIT]
Looks like it wasn't working because I'd set $PATH at the front instead of being last
Wrong: `ANDROID_NDK_ROOT=<dir>/23.1.7779620 PATH="$PATH:$ANDROID_NDK_ROOT/toolchains/llvm/prebuilt/linux-x86_64/bin" ../Configure android-arm -D_ANDROID_API_=21`
Correct: `ANDROID_NDK_ROOT=<dir>/23.1.7779620 PATH="$ANDROID_NDK_ROOT/toolchains/llvm/prebuilt/linux-x86_64/bin:$PATH" ../Configure android-arm -D_ANDROID_API_=21`

Please disregard all instructions above the edit
